pkg/admission/server.go

```diff
@@ -0,0 +1,157 @@
+	// Client will be injected by the manager if not set.
+	Client client.Client
+
+	// BootstrapOptions contains the options for bootstrapping the admission server.
+	*BootstrapOptions
+}
+
+// BootstrapOptions are options for bootstrapping an admission webhook server.
+type BootstrapOptions struct {
+	// Secret is the location for storing the certificate for the admission server.
+	// The server should be permission to create a secret in the namespace.
+	// This is optional. If unspecified, it will write to the filesystem.
+	Secret *types.NamespacedName
+
+	// Host is the host name of .webhooks.clientConfig.url
+	// https://github.com/kubernetes/api/blob/183f3326a9353bd6d41430fc80f96259331d029c/admissionregistration/v1beta1/types.go#L250
+	// If unspecified, will be defaulted to "localhost".
+	Host *string
+
+	// Service is k8s service fronting the webhook server pod(s).
+	// This field is optional.
+	// This maps to field .webhooks.clientConfig.service
+	// https://github.com/kubernetes/api/blob/183f3326a9353bd6d41430fc80f96259331d029c/admissionregistration/v1beta1/types.go#L260
+	Service *types.NamespacedName
+	// Labels is the selector of the service
+	Labels map[string]string
+
+	// CertProvisioner provisions certificates for the admission webhook server.
+	// This is optional. If unspecified, it will be defaulted to use a self-signed certificate.
+	CertProvisioner *controller.CertProvisioner
+}
+
+// Server is an admission webhook server that can serve traffic and
+// generates related k8s reources for deploying.
+type Server struct {
+	// Name is the name of server
+	Name string
+
+	sMux *http.ServeMux
+	// registry maps a path to a httpHandler.
+	registry map[string]httpHandler
+
+	// ServerOptions contains options for configuring the admission server.
+	ServerOptions
+
+	once sync.Once
+}
+
+// NewServer creates a new admission webhook server.
+func NewServer(name string, mgr manager.Manager, ops ServerOptions) (*Server, error) {
+	as := &Server{
+		Name:          name,
+		ServerOptions: ops,
+	}
+
+	return as, mgr.Add(as)
+}
+
+// DefaultServer returns an Server and defaults all possible fields.
+var DefaultServer = &Server{}
+
+// Register registers webhook(s) in the server
+func (s *Server) Register(webhooks ...*Webhook) error {
+	for _, webhook := range webhooks {
+		_, found := s.registry[webhook.path]
+		if found {
+			return fmt.Errorf("can't register duplicate path: %v", webhook.path)
+		}
+		httpHandler := httpHandler{
+			Webhook:     webhook,
+			Annotations: s.Annotations,
+		}
+		s.registry[webhook.path] = httpHandler
+		s.sMux.Handle(webhook.path, httpHandler)
+	}
+	return nil
+}
+
+var _ manager.Runnable = &Server{}
+
+// Start runs the server.
+func (s *Server) Start(stop <-chan struct{}) error {
+	_, err := s.bootstrap(false)
+	if err != nil {
+		return err
+	}
+	// TODO(mengqiy): add jitter to the timer
+	//timer := time.Tick(6 * 30 * 24 * time.Hour)
+	errCh := make(chan error)
+	go s.listenAndServeTLS(errCh)
```

Comment:
This will block the `Start(stopCh)`. Most of the time, `Start` implies non-blocking. also `stop` doesn't stop the server.